include/seqan3/utility/views/zip.hpp

```diff
@@ -6,26 +6,476 @@
+            std::conditional_t<all_random_access<Const, Views...>, std::random_access_iterator_tag,
+                std::conditional_t<all_bidirectional<Const, Views...>, std::bidirectional_iterator_tag,
+                    std::conditional_t<all_forward<Const, Views...>, std::forward_iterator_tag,
+                        std::input_iterator_tag>>>>;
+};
+#else // cpp20 iterators
+template <bool> struct iterator_category_t;
+template <> struct iterator_category_t<true> { using iterator_category = std::input_iterator_tag; };
+template <> struct iterator_category_t<false> {};
+#endif
+
+} // namespace seqan3::detail::zip
+
+template <std::ranges::input_range... Views>
+    requires (std::ranges::view<Views> && ...) && (sizeof...(Views) > 0)
+template <bool Const>
+#if defined(__GNUC__) && (__GNUC__ == 10) // cpp17 iterators
+class zip_view<Views...>::iterator : public zip::iterator_category_t<Const, Views...>
+#else // cpp20 iterators
+class zip_view<Views...>::iterator : public zip::iterator_category_t<zip::all_forward<Const, Views...>>
+#endif
+{
+private:
+    constexpr explicit iterator(
+        zip::tuple_or_pair<std::ranges::iterator_t<view_helper::maybe_const<Const, Views>>...> current)
+    : current_(std::move(current))
+    {}
+
+    friend class zip_view<Views...>;
+
+public:
+    // Exposition-only. Is public for access via friend operator== of the sentinel.
+    zip::tuple_or_pair<std::ranges::iterator_t<view_helper::maybe_const<Const, Views>>...> current_;
+
+    using iterator_concept =
+        std::conditional_t<zip::all_random_access<Const, Views...>, std::random_access_iterator_tag,
+            std::conditional_t<zip::all_bidirectional<Const, Views...>, std::bidirectional_iterator_tag,
+                std::conditional_t<zip::all_forward<Const, Views...>, std::forward_iterator_tag,
+                    std::input_iterator_tag>>>;
+    using value_type = zip::tuple_or_pair<std::ranges::range_value_t<view_helper::maybe_const<Const, Views>>...>;
+    using difference_type = std::common_type_t<
+                                std::ranges::range_difference_t<view_helper::maybe_const<Const, Views>>...>;
+
+    iterator() = default;
+    constexpr iterator(iterator<!Const> i)
+        requires Const && (std::convertible_to<std::ranges::iterator_t<Views>,
+                                               std::ranges::iterator_t<view_helper::maybe_const<Const, Views>>> && ...)
+    : current_(std::move(i.current))
+    {}
+
+    constexpr auto operator*() const
+    {
+        return zip::tuple_transform([] (auto & i) -> decltype(auto) { return *i; }, current_);
+    }
+    constexpr iterator& operator++()
+    {
+        zip::tuple_for_each([] (auto & i) { ++i; }, current_);
+        return *this;
+    }
+    constexpr void operator++(int)
+    {
+        ++*this;
+    }
+    constexpr iterator operator++(int)
+        requires zip::all_forward<Const, Views...>
```

Comment:
If it is not `all_forward` we can't copy?

Reply:
With pre-cpp20 iterators you were able to copy input iterators, but once you incremented one of them, you had no guarantees (https://en.cppreference.com/w/cpp/named_req/InputIterator). Cpp20 input iterators are now move only.

Reply:
ah ok. thanks for clarifying. I don't think our other views are conform with this?